### PR TITLE
Building intellij-pants-plugin from sources

### PR DESCRIPTION
**Pull request by odisseus**
It takes about 16 minutes till pants are installed.

### CODE REVIEW

File: driver/src/main/scala/org/virtuslab/ideprobe/pants/PantsPluginBuilder.scala [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+    val deploy = List("./scripts/deploy/deploy.sh", "--skip-publish")
+    val result = Shell.run(localRepo, env, deploy: _*)
+    if (result.exitCode != 0) throw new Exception(s"Couldn't build pants plugin. STDERR:\n${result.err}")
+
+    val files = localRepo.toFile.listFiles()
+    val pantsPath = files.find(_.getName.matches("pants_.*\\.zip")).getOrElse {
+      throw new Exception(s"Couldn't find pants archive. Existing files:\n${files.mkString("\n")}")
+    }
+    println(s"Built pants at $pantsPath")
+    pantsPath.toPath.inputStream
```
Comment (lukaszwawrzyk) [inline driver/src/main/scala/org/virtuslab/ideprobe/pants/PantsPluginBuilder.scala:42]: you may want to move the file to some tmp location and remove cloned repository

File: driver/src/main/scala/org/virtuslab/ideprobe/util/GitUtils.scala [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+
+import java.nio.file.Files
+import java.nio.file.Path
+
+import org.virtuslab.ideprobe.Shell
+import org.virtuslab.ideprobe.dependencies.SourceRepository.Git
+
+object GitUtils {
+  def clone(repository: Git): Path = {
+    val localRepo = Files.createTempDirectory("scala-plugin-repo")
```
Comment (lukaszwawrzyk) [inline driver/src/main/scala/org/virtuslab/ideprobe/util/GitUtils.scala:11]: maybe change the name ;p

Comment (odisseus) [reply driver/src/main/scala/org/virtuslab/ideprobe/util/GitUtils.scala:11]: oops

File: driver/src/main/scala/org/virtuslab/ideprobe/pants/PantsPluginBuilder.scala (hunk 1 above)
Comment (odisseus) [reply driver/src/main/scala/org/virtuslab/ideprobe/pants/PantsPluginBuilder.scala:42]: I don't think that's necessary; [deleting .NET and Android stuff](https://github.com/VirtusLab/intellij-pants-plugin-tests/pull/8) yielded plenty of space.

File: driver/src/main/scala/org/virtuslab/ideprobe/pants/PantsPluginBuilder.scala [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+        val artifact = git.path.resolveChild(hash)
+
+        resources.get(artifact, provider = () => build(git))
+    }
+
+  private def build(repository: Git): InputStream = {
+    val localRepo = GitUtils.clone(repository)
+
+    val setupCiEnvironment = List("./scripts/setup-ci-environment.sh")
+    val result0 = Shell.run(localRepo, Map.empty[String, String], setupCiEnvironment: _*)
```
[the remark this answers is not in the document]
Comment (odisseus) [reply driver/src/main/scala/org/virtuslab/ideprobe/pants/PantsPluginBuilder.scala:29]: Refactored

**Comment by lukaszwawrzyk** [comment]
Really nice we can use pants from source now. Currently we added some notable cost to each build. I think next step is to cleanup and optimize docker images.
* move installing zip to base image
* add time measurement for actions like building plugin, cloning repo in pants etc. so we clearly see what costs how much
* inspect base image - essentially we want to have there things that take time to set up but don't change frequently
The fake test at: com.twitter.intellij.pants.PreparePants is run inside that container that sets up a fixture, installs intellij and plugins (so also builds plugin) and sets up pants from sources. It is worth using it. Basically, if you just generate the image in this way (install all deps and run the test), and use the image to run other tests, the tests should start up instantly and run quickly (no need to download or build anything). It is worth to try this as sanity check. Our final goal is to build such image in our CI. So with that said:
- Step 1. Build image that has all deps and has run the PreparePants and test if running tests from this image doesn't trigger unncesary long running actions
- Step 2. Optimize the base image: inspect all files created there and delete what is not necessary. For example I would delete all cloned repositories etc. I am also considering deleting intellij, this would save us rebuilding image on each intellij version update. For sure we want to keep rust cache, also coursier cache, sbt. Make sure that cleanup is done in the same layer as building to ensure the files don't land in any layer of final image increasing its size.
- Step 3. Upload the optimized image to dockerhub
- Step 4. Just run tests using the base image in separate jobs and see how it goes

Note: The idea I had to build a second image based on 'base image' with built plugin and fetched intellij, etc. per build sounds useless. We would have one worker, that, say, would work for 5 min to prepare image, upload it, and only then other jobs would start, download the image and run tests immedietly. But instead waiting 5 min, jobs could just just build the image themselves in 5 mins and skip the time for upload and download. Hopefully this time should be short, much shorter than now, with a proper base image.

**Review by lukaszwawrzyk — APPROVED** [review]

**Pull request merged**